maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -847,12 +847,33 @@ private DefaultScanResult scanDependencies()
         {
             try
             {
+                DefaultScanResult scanResult = new DefaultScanResult( Collections.EMPTY_LIST );
+                
+                List<String> dependenciesToScanList = new ArrayList( Arrays.asList( getDependenciesToScan() ) );
+                TestListResolver includedAndExcludedTests = getIncludedAndExcludedTests();
+
+                for ( MavenProject mavenProject : session.getSortedProjects() )
+                {
+                    String groupArtifactId = new StringBuilder( mavenProject.getGroupId() ).append( ":" )
+                            .append( mavenProject.getArtifactId() ).toString();
+                    if ( dependenciesToScanList.removeAll( Collections.singleton( groupArtifactId ) ) )
+                    {
+                        File outputDirectoryFile = new File( mavenProject.getBuild().getOutputDirectory() );
+                        DirectoryScanner scanner =
+                                new DirectoryScanner( outputDirectoryFile, includedAndExcludedTests );
+                        scanResult = scanResult.append( scanner.scan() );
+                    }
+                    
+                }
+
                 // @TODO noinspection unchecked, check MavenProject 3.x for Generics in surefire:3.0
                 @SuppressWarnings( "unchecked" )
-                List<File> dependenciesToScan =
-                    DependencyScanner.filter( project.getTestArtifacts(), Arrays.asList( getDependenciesToScan() ) );
-                DependencyScanner scanner = new DependencyScanner( dependenciesToScan, getIncludedAndExcludedTests() );
-                return scanner.scan();
+                List<File> dependenciesToScanFileList =
```

Comment:
`dependenciesToScanFileList` already contains plural. Pls rename `dependenciesToScanFileList` to `dependenciesToScanFile`.